### PR TITLE
Port 3.3 changes to 3.5

### PR DESCRIPTION
**Pull request by poornachandra**
JIRAs:
https://issues.cask.co/browse/CDAP-6860 (3.3 PR https://github.com/caskdata/cdap/pull/6280)
https://issues.cask.co/browse/CDAP-6852 (3.3 PR https://github.com/caskdata/cdap/pull/6272)
https://issues.cask.co/browse/CDAP-6829 (3.3 PR https://github.com/caskdata/cdap/pull/6251)

Build - http://builds.cask.co/browse/CDAP-RUT11-1

Note: Except for commit `CDAP-6829 Handle non-existent file during cleanup` (https://github.com/caskdata/cdap/pull/6321/commits/2bbcb00b052ab1aeee955cbca02db6858ba195b3), all other commits are cherry-picks without any conflicts.

### CODE REVIEW

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/write/LogCleanupTest.java [hunk 1]
```diff
@@ -91,9 +100,24 @@ public static void init() throws Exception {
...
+      Modules.override(new LocationRuntimeModule().getDistributedModules()).with(
+        new AbstractModule() {
+          @Override
+          protected void configure() {
+            // Override namespace location factory so that it does not perform lookup for NamespaceMeta like
+            // DefaultNamespacedLocationFactory does and hence allow unit
+            // tests to use it without creating namespace meta for the namespace.
+            // This is similar to what NonCustomLocationUnitTestModule does.
+            bind(NamespacedLocationFactory.class).to(NamespacedLocationFactoryTestClient.class);
+            bind(LocationFactory.class).toInstance(localLocationFactory);
```
Comment (rohitsinha54) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/write/LogCleanupTest.java]: Might be useful to put a comment that we need this to make LocationFactory use the dir for this junit test.

File: cdap-watchdog/src/test/java/co/cask/cdap/logging/write/LogCleanupTest.java [hunk 2]
```diff
@@ -261,7 +299,7 @@ public void testDeleteEmptyDir1() throws Exception {
   @Test
   public void testDeleteEmptyDir2() throws Exception {
     // Create base dir
-    LocationFactory locationFactory = injector.getInstance(LocationFactory.class);
+    LocationFactory locationFactory = injector.getInstance(LocalLocationFactory.class);
```
Comment (rohitsinha54) [inline cdap-watchdog/src/test/java/co/cask/cdap/logging/write/LogCleanupTest.java]: Why change this LocationFactory got binded the the instance of LocalLocationFactory created above no ?

**Comment by rohitsinha54** [comment]
Very minor comments other than that LGTM assuming all other commits except  
CDAP-6829 Handle non-existent file during cleanup are direct cherry-pick.

**Comment by rohitsinha54** [comment]
ITM: http://builds.cask.co/browse/CDAP-ITM0-4

**Comment by poornachandra** [comment]
@rohitsinha54 I've addressed your comments, please take a look.

**Comment by rohitsinha54** [comment]
LGTM

**Comment by poornachandra** [comment]
Build is green, merging...

**Pull request merged**